logical/framework/openapi.go

```diff
@@ -0,0 +1,548 @@
+	// detect and remove it now than to compensate for in the other regexes.
+	//
+	// example: (?P<foo>\\w(([\\w-.]+)?\\w)?) -> (?P<foo>)
+	base := GenericNameRegex("")
+	start := strings.Index(base, ">")
+	end := strings.LastIndex(base, ")")
+	regexToRemove := ""
+	if start != -1 && end != -1 && end > start {
+		regexToRemove = base[start+1 : end]
+	}
+
+	pattern = strings.Replace(pattern, regexToRemove, "", -1)
+
+	// Initialize paths with the original pattern or the halves of an
+	// alternation, which is also present in some patterns.
+	matches := altRe.FindAllStringSubmatch(pattern, -1)
+	if len(matches) > 0 {
+		paths = []string{matches[0][1], matches[0][2]}
+	} else {
+		paths = []string{pattern}
+	}
+
+	// Expand all optional regex elements into two paths. This approach is really only useful up to 2 optional
+	// groups, but we probably don't want to deal with the exponential increase beyond that anyway.
+	for i := 0; i < len(paths); i++ {
+		p := paths[i]
+
+		// match is a 2-element slice that will have a start and end index
+		// for the left-most match of a regex of form: (lease/)?
+		match := optRe.FindStringIndex(p)
+
+		if match != nil {
+			// create a path that includes the optional element but without
+			// parenthesis or the '?' character.
+			paths[i] = p[:match[0]] + p[match[0]+1:match[1]-2] + p[match[1]:]
+
+			// create a path that excludes the optional element.
+			paths = append(paths, p[:match[0]]+p[match[1]:])
+			i--
+		}
+	}
+
+	// Replace named parameters (?P<foo>) with {foo}
+	var replacedPaths []string
+
+	for _, path := range paths {
+		result := reqdRe.FindAllStringSubmatch(path, -1)
+		if result != nil {
+			for _, p := range result {
+				par := p[1]
+				path = strings.Replace(path, p[0], fmt.Sprintf("{%s}", par), 1)
+			}
+		}
+		// Final cleanup
+		path = cleanSuffixRe.ReplaceAllString(path, "")
+		path = cleanCharsRe.ReplaceAllString(path, "")
+		replacedPaths = append(replacedPaths, path)
+	}
+
+	return replacedPaths
+}
+
+// schemaType is a subset of the JSON Schema elements used as a target
+// for conversions from Vault's standard FieldTypes.
+type schemaType struct {
+	baseType string
+	items    string
+	format   string
+}
+
+// convertType translates a FieldType into an OpenAPI type.
+// In the case of arrays, a subtype is returned as well.
+func convertType(t FieldType) schemaType {
+	ret := schemaType{}
+
+	switch t {
+	case TypeString, TypeNameString, TypeLowerCaseString, TypeHeader:
+		ret.baseType = "string"
+	case TypeInt:
+		ret.baseType = "number"
+	case TypeDurationSecond:
+		ret.baseType = "number"
+		ret.format = "seconds"
+	case TypeBool:
+		ret.baseType = "boolean"
+	case TypeMap:
+		ret.baseType = "object"
+		ret.format = "map"
+	case TypeKVPairs:
+		ret.baseType = "object"
+		ret.format = "kvpairs"
+	case TypeSlice, TypeStringSlice, TypeCommaStringSlice:
+		ret.baseType = "array"
+		ret.items = "string"
```

Comment:
This is not always a string. TypeSlice can be any type.

Reply:
Updated so TypeSlice items will be "object".